src/lib/utils/ct_utils.h

```diff
@@ -173,6 +173,63 @@ class Choice final {
       uint32_t m_value;
 };
 
+/**
+* A CT::Option<T> is either a valid T (Some), or a default T (None)
+*
+* T must be default construtible
+*/
+template <typename T>
+class Option {
+   public:
+      constexpr Option(T v, Choice valid) : m_is_some(valid), m_value(v) {}
+
+      constexpr Option(T v) : m_is_some(Choice::yes()), m_value(v) {}
+
+      constexpr Option() : m_is_some(Choice::no()), m_value(T()) {}
+
+      /// Return if this Option is Some
+      constexpr Choice is_some() const { return m_is_some; }
+
+      /// Return if this Option is None
+      constexpr Choice is_none() const { return !m_is_some; }
+
+      /**
+      * Apply a function to the inner value and return a new Option
+      * which contains that value. This is constant time only if `f` is.
+      */
+      template <typename F>
+      constexpr auto map(F f) const {
+         typedef typename std::invoke_result<F, T>::type U;
+         return Option<U>(f(m_value), m_is_some);
+      }
+
+      /// Either returns the value or throws an exception
+      T unwrap() const {
+         BOTAN_STATE_CHECK(m_is_some.as_bool());
+         return m_value;
+      }
+
+      /// Returns either the inner value or the alternative, in constant time
+      T unwrap_or(const T& other) const {
+         T r = other;
+         r.conditional_assign(m_is_some, m_value);
+         return r;
+      }
```

Comment:
Probably a good idea to scope this to types that actually provide this method. And also make the copy of `other` visible via the interface (to optimize usage with rvalue refs for instance):

```suggestion
      /// Returns either the inner value or the alternative, in constant time
      T unwrap_or(T other) const 
         requires concepts::ct_conditional_assignable<T>
      {
         other.conditional_assign(m_is_some, m_value);
         return other;
      }
```

... and the concept itself:

```C++
namespace concepts {

template <typename T>
concept ct_conditional_assignable = requires(T lhs, const T& rhs, Choice c) { lhs.conditional_assign(c, rhs); };

}  // namespace concepts

}
```

... if you wanted to, you could even provide a variable time overload if the concept isn't fulfilled. Though, I'd consider this fairly reckless, as it would potentially fool the user if their type doesn't model the concept (perhaps even subtly inadvertantly).

Reply:
I'd be quite against a variable time overload. That said it would be nice if we could handle this more generically, ie `CT::Option<uint32_t>` is entirely a sensible thing, and it's quite possible to implement `unwrap_or` in constant time as well, but requires additional help. I'll ponder this.

Reply:
Nice way of expressing the conditional assignment btw.

I made a change such that `value_or` can handle either types which explicitly provided an appropriate `conditional_assign`, and for unsigned integral types which make use of `CT::Mask` to perform the selection.